src/compiler/crystal/command/format.cr

```diff
@@ -69,27 +75,24 @@ class Crystal::Command
       files = normalize_paths files
     end
 
-    format_many files, check_files, excludes
-
-    if check_files
-      if check_files.empty?
-        exit 0
-      else
-        check_files.each do |result|
-          case result.code
-          when .format?
-            error "formatting '#{result.filename}' produced changes", exit_code: nil
-          when .syntax?
-            error "'#{result.filename}' has syntax errors", exit_code: nil
-          when .invalid_byte_sequence?
-            error "'#{result.filename}' is not a valid Crystal source file", exit_code: nil
-          when .bug?
-            error "there's a bug formatting '#{result.filename}', to show more information, please run:\n\n  $ crystal tool format '#{result.filename}'", exit_code: nil
-          end
+    format_many files, check, check_files, excludes, show_backtrace
+
+    if check
+      check_files.each do |result|
+        case result.code
+        when .format?
+          error "formatting '#{result.filename}' produced changes", exit_code: nil
+        when .syntax?
+          error "'#{result.filename}' has syntax errors", exit_code: nil
+        when .invalid_byte_sequence?
+          error "'#{result.filename}' is not a valid Crystal source file", exit_code: nil
+        when .bug?
+          error "there's a bug formatting '#{result.filename}'", exit_code: nil
         end
-        exit 1
       end
     end
+
+    exit check_files.empty? ? 0 : 1
```

Comment:
So now if check is false, and there are files with syntax error or invalid byte sequence the formatter will return a non zero status.

I am not sure that an element should be appended if `check == false` when handling `ex : InvalidByteSequenceError` and `ex : Crystal::SyntaxException`.

When the user is half way editing a file and the editor use the formatter I think we want that to succeed (even if it is unable to do some format)

Reply:
`SyntaxException` and `InvalidByteSequenceError` are also error. So, it should report as error generally. But they has difference than a bug. They are expected on formatter, and a bug is unexpected.
It may be needed to distinguish both a bug (unexpected error) and other errors (expected error), so `crystal tool format` returns `1` on expected errors and `2` on unexpected errors. And, editor plugin applying formatter checks status code of `crystal tool format`, then it shows pop-up only if the status code is `2`, but ignores errors when status code is `1` because it is not useful on editor plugin.

Reply:
@bcardiff Currently, the formatter returns `1` in these cases. It doesn't seem to be an issue.

On the contrary, changing that would probably break things.

Reply:
I'm fine with signaling the type of error by different return codes. But I wouldn't bother unless there is a real use case for this.

Reply:
Ok, I guess I misunderstood the proposed changes regarding the exit status.